osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/firmware/UpdateFirmwareCommandExecutor.java

```diff
@@ -158,30 +157,30 @@ private FirmwareFile addMac(
       final FirmwareFile firmwareFile)
       throws ProtocolAdapterException {
 
-    log.debug("Adding MAC to firmware file for M-Bus device");
+    log.debug(
+        "Adding MAC to firmware file for M-Bus device with deviceIdentification {}",
+        deviceIdentification);
     final DlmsDevice mbusDevice =
         this.dlmsDeviceRepository.findByDeviceIdentification(deviceIdentification);
     if (mbusDevice == null) {
       throw new ProtocolAdapterException(
           String.format(EXCEPTION_MSG_DEVICE_NOT_AVAILABLE_IN_DATABASE, deviceIdentification));
     }
 
-    final Long hexNotationOfTheMbusIdentificationNumber = mbusDevice.getMbusIdentificationNumber();
-    if (hexNotationOfTheMbusIdentificationNumber == null) {
+    final Long mbusIdentificationNumberFromDatabase = mbusDevice.getMbusIdentificationNumber();
+    if (mbusIdentificationNumberFromDatabase == null) {
       throw new ProtocolAdapterException(
           String.format(
               EXCEPTION_MSG_DEVICE_HAS_NO_MBUS_IDENTIFICATION_NUMBER, deviceIdentification));
     }
 
-    final Long mbusIdentificationNumber =
-        new IdentificationNumber(String.format("%08d", hexNotationOfTheMbusIdentificationNumber))
-            .getIdentificationNumber();
+    // The MbusIdentificationNumber from the osgp_adapter_protocol_dlms database
+    // is used directly as int value in FirmwareFile header and subsequently the ImageIdentifier.
+    // This only works for SMR5+ devices
```

Comment:
This may have to do for now, in order to be able to get some progress fixing the M-Bus firmware update.
This will have to be re-assessed, and possibly changed, if we have more clarity about how the identification number is to be treated.

Plainly stating to use the value "from the database" will need to be revisited as well (possibly not immediately).
"From the database" does not tell anything about the meaning/treatment in DLMS/DSMR/SMR and is subject to potential change that should not affect the functionality here unnoticed. The way the data are stored in the DB should be an implementation detail.

Reply:
There is a difference in th specification of the **_M-Bus Client Setup_** (Class ID: 72, OBIS-code: 0-x:24.1.0.255) between DSMR4 and SMR5
In **DSMR4.2** the _M-Bus Client Setup_ has **version 0** and the _identification_number_ attribute is defined as double-long unsigned **interpreted as BCD** where in **SMR5** the _M-Bus Client Setup_ has **version 1** and the _identification_number_ attribute is defined as double-long unsigned
So the value from the database is undefined. It depends on the version in the DlmsObjectConfig how to interpret this value. The conversion of the database value to the Numerical or Textual representation is handled by the IdentificationNumber object.

```
  private IdentificationNumber getIdentificationNumber(final DlmsDevice mbusDevice)
      throws ProtocolAdapterException {
    final Long mbusIdentificationNumberFromDatabase = mbusDevice.getMbusIdentificationNumber();

    final IdentificationNumber identificationNumber;

    if (this.mbusIdentificationIsBcdRepresentatyionAsLong(mbusDevice)) {

      identificationNumber =
          IdentificationNumber.fromBcdRepresentationAsLong(mbusIdentificationNumberFromDatabase);

    } else {
      identificationNumber =
          IdentificationNumber.fromNumericalRepresentation(mbusIdentificationNumberFromDatabase);
    }
    return identificationNumber;
  }

  private boolean mbusIdentificationIsBcdRepresentatyionAsLong(final DlmsDevice mbusDevice)
      throws ProtocolAdapterException {

    final DlmsObject mbusClientSetupObject =
        this.dlmsObjectConfigService.getDlmsObject(mbusDevice, DlmsObjectType.MBUS_CLIENT_SETUP);

    return mbusClientSetupObject.getVersion().equals(DlmsClassVersion.VERSION_0);
  }
```